src/form/df.jl

```diff
@@ -0,0 +1,163 @@
+# this file contains (balanced) convexified DistFlow formulation, in W space
+export
+    SOCDFPowerModel, SOCDFForm
+""
+abstract type AbstractDFForm <: AbstractPowerFormulation end
+
+""
+abstract type SOCDFForm <: AbstractDFForm end
+
+""
+const SOCDFPowerModel = GenericPowerModel{SOCDFForm}
+
+"default SOC constructor"
+SOCDFPowerModel(data::Dict{String,Any}; kwargs...) = GenericPowerModel(data, SOCDFForm; kwargs...)
+
+""
+function variable_branch_current(pm::GenericPowerModel{T}, n::Int=pm.cnw; kwargs...) where T <: AbstractDFForm
+    variable_branch_series_current_magnitude_sqr(pm, n; kwargs...)
+end
+
+""
+function variable_branch_flow(pm::GenericPowerModel{T}, n::Int=pm.cnw; kwargs...) where T <: AbstractDFForm
+    variable_active_branch_flow(pm, n; kwargs...)
+    variable_reactive_branch_flow(pm, n; kwargs...)
+end
+
+""
+function variable_voltage(pm::GenericPowerModel{T}, n::Int=pm.cnw; kwargs...) where T <: AbstractDFForm
+    variable_voltage_magnitude_sqr(pm, n; kwargs...)
+end
+
+"do nothing, this model does not have complex voltage constraints"
+function constraint_voltage(pm::GenericPowerModel{T}, n::Int) where T <: AbstractDFForm
+end
+
+"""
+Defines branch flow model power flow equations
+"""
+function constraint_flow_losses(pm::GenericPowerModel{T}, n::Int, i, f_bus, t_bus, f_idx, t_idx, r, x, g_sh_fr, g_sh_to, b_sh_fr, b_sh_to, tm) where T <: AbstractDFForm
+    p_fr = pm.var[:nw][n][:p][f_idx]
+    q_fr = pm.var[:nw][n][:q][f_idx]
+    p_to = pm.var[:nw][n][:p][t_idx]
+    q_to = pm.var[:nw][n][:q][t_idx]
+    w_fr = pm.var[:nw][n][:w][f_bus]
+    w_to = pm.var[:nw][n][:w][t_bus]
+    ccm =  pm.var[:nw][n][:ccm][i]
+
+    @constraint(pm.model, p_fr + p_to ==  g_sh_fr*(w_fr/tm^2) + r*ccm +  g_sh_to*w_to)
+    @constraint(pm.model, q_fr + q_to == -b_sh_fr*(w_fr/tm^2) + x*ccm + -b_sh_to*w_to)
+end
+
+
+"""
+Defines voltage drop over a branch, linking from and to side voltage magnitude
+"""
+function constraint_voltage_magnitude_difference(pm::GenericPowerModel{T}, n::Int, i, f_bus, t_bus, f_idx, t_idx, r, x, g_sh_fr, b_sh_fr, tm) where T <: AbstractDFForm
+    p_fr = pm.var[:nw][n][:p][f_idx]
+    q_fr = pm.var[:nw][n][:q][f_idx]
+    w_fr = pm.var[:nw][n][:w][f_bus]
+    w_to = pm.var[:nw][n][:w][t_bus]
+    ccm =  pm.var[:nw][n][:ccm][i]
+
+    #define series flow expressions to simplify Ohm's law
+    p_fr_s = p_fr - g_sh_fr*(w_fr/tm^2)
+    q_fr_s = q_fr + b_sh_fr*(w_fr/tm^2)
```

Comment:
It's not immediately obvious to me that this is correct.  Did you double check?  Until we have an asymmetrical network test this will not be checked by unit tests.

Reply:
The derivation goes as follows. Branches are modeled with an ideal transformer and a pi-section in series.

As the ideal transformer is located at the from side, the voltage after the transformer, i.e. at the from side of the pi-element following it, is actually `(w_fr/tm^2)`. Furthermore, as this transformer is ideal, it is lossless. Moreover, as we'll be doing the DistFlow derivation, which performs the angle relaxation, transformer phase shift does not influence the power flow (except for `constraint_voltage_angle_difference`.)

We observe that a a branch's power flow balance with asymmetrical shunts is composed of three loss contributions: `p_fr + p_to = p_sh_loss_fr + p_s_loss + p_sh_loss_to`, `q_fr + q_to = q_sh_loss_fr + q_s_loss + q_sh_loss_to`, where `x_sh_loss_fr`, `x_sh_loss_to` are the to from and to side shunt losses, and `x_s_loss` is the series flow loss. We can further define the flow only through the series impedance as `p_fr_s + p_to_s = p_s_loss`, `q_fr_s + q_to_s = q_s_loss`. 

The pi-section loss terms, expressed in the `w` and `ccm` variables therefore are:
- `p_sh_loss_fr =  g_sh_fr*(w_fr/tm^2)`
- `q_sh_loss_fr = -b_sh_fr*(w_fr/tm^2)`
- `p_s = r*ccm` 
- `p_s = x*ccm`
- `p_sh_loss_to =  g_sh_to*(w_to)`
- `q_sh_loss_to = -b_sh_to*(w_to)`

Ohm's law only relates the voltage magnitude and the series current and series impedance of the pi-element. Therefore, after performing the DistFlow derivation (`v_j = v_i - z_ij *I_ij_s`, with RHS and LHS multiplied by their complex conjugated) and the subsititution of the squared node voltage magnitude, (with `w` and squared branch *series* current magnitude, with `ccm`), you obtain expressions in the *series* complex power flow variables `p_fr_s, q_fr_s`. Furthermore, in these variables `p_fr_s^2 +q_fr_s^2 == (w_fr/tm^2)*ccm` is a valid expression, which is then relaxed. 

I can add this somewhere in the documentation.

I haven't found the time to rederive `constraint_voltage_angle_difference` to confirm its current implementation though.